Add Jupyter Server Architecture diagram (#801)
jupyter-server/jupyter_server: docs/source/developers/architecture.rst

.. _architecture:

Architecture Diagrams
=====================

This page describes the Jupyter Server architecture and the main workflows.
This information is useful for developers who want to understand how Jupyter
Server components are connected and how the principal workflows look like.

To make changes for these diagrams, use `the Draw.io <https://app.diagrams.net/>`_

> open source tool.

Reviewer: ```suggestion
open source tool to edit the png file.
```

Jupyter Server Architecture
---------------------------

> The Jupyter Server system can be seen in figure bellow:

Reviewer: ```suggestion
The Jupyter Server system can be seen in figure below:
```

.. image:: ../images/jupyter-server-architecture.drawio.png
   :alt: Jupyter Server Architecture
   :width: 100%
   :align: center

Jupyter Server contains the following components:

- **ServerApp** is the main Tornado-based application which connects all
  components together.

- **Config Manager** initializes configuration for the ServerApp. You can define
  custom classes for the Jupyter Server managers using this config and change
  SererApp settings. Follow :ref:`the Config File Guide <other-full-config>` to
  learn about configuration settings and how to build custom config.

> - **Custom Extensions** allow to create the custom Server's REST API endpoint.

Reviewer: ```suggestion
- **Custom Extensions** allow you to create the custom Server's REST API endpoint.
```

  Follow :ref:`the Extension Guide <extensions>` to know more about extending
  ServerApp with extra request handlers.

> - **Kernel Gateway** is the web server that provides access to Jupyter Kernels.

Reviewer: ```suggestion
- **Kernel Gateway** is the web server that provides access to remote Jupyter Kernels.
```

> - **Kernel Gateway** is the web server that provides access to Jupyter Kernels.
>   There are different ways to create this gateway. If your ServerApp needs to
>   communicate with remote Kernels, you can use
>   `the Enterprise Gateway <https://github.com/jupyter-server/enterprise_gateway>`_,
>   otherwise you can use `the Kernel Gateway <https://github.com/jupyter-server/kernel_gateway>`_.

Reviewer 2: Because the box states, `Kernel Gateway` yet both forms of gateway servers are listed here, perhaps using the term `Gateway Server` may be more clear:
```suggestion
- **Gateway Server** is a web server that, when configured, provides access to Jupyter Kernels running on other hosts.
  There are different ways to create a gateway server. If your ServerApp needs to
  communicate with remote Kernels residing within resource-managed clusters, you can use
  `Enterprise Gateway <https://github.com/jupyter-server/enterprise_gateway>`_,
  otherwise, you can use `Kernel Gateway <https://github.com/jupyter-server/kernel_gateway>`_, where Kernels run locally to the gateway server.
```

Author: I agree with that, I will rename `Kernel Gateway` to `Gateway Server` on the diagram also.

- **Contents Manager and File Contents Manager** are responsible for serving
  Notebook on the file system. Session Manager uses Contents Manager to receive
  Kernel path. Follow :ref:`the Contents API guide <contents_api>` to learn
  about Contents Manager.

> - **Session Manager** processes users Sessions. When user starts a new Kernel,

Reviewer: ```suggestion
- **Session Manager** processes users' Sessions. When a user starts a new Kernel,
```

  Session Manager starts a process to provision Kernel for the user and generates
  a new Session ID. Each opened Notebook has a separate Session, but different

>   Notebook Kernels can use the same Session. That is useful if user wants to

Reviewer: ```suggestion
  Notebook Kernels can use the same Session. That is useful if the user wants to
```

  share data across various opened Notebooks. Session Manager uses SQLite3

>   DataBase to store the Sessions.

Reviewer: ```suggestion
  DataBase to store the Session information.  The database is stored in memory by default, but can be configured to save to disk.
```

> - **Mapping Kernel Manager** is responsible to operate multiple Kernels in the
>   ServerApp. It starts a new Kernel for a user's Session and generates a
>   new Kernel ID.

Reviewer 2: ```suggestion
- **Mapping Kernel Manager** is responsible for managing the lifecycles of the Kernels running within the
  ServerApp. It starts a new Kernel for a user's Session and facilitates interrupt, restart, and shutdown operations against the kernel.
```

> - **Kernel Spec Manager** parses file with JSON specification for a Kernel.

Reviewer: ```suggestion
- **Kernel Spec Manager** parses files with JSON specification for Kernels, and provides a list of available kernel configurations.
```

Reviewer 2: I find the `Jupyter Client` reference in the diagram to be a little out of place.  I wonder if we were to also add `Kernel Manger` with a relationship between it and `MappingKernelManager` it might help clarify.  If so, I would suggest adding an entry labeled **Jupyter Client** with sub-entries of **Kernel Spec Manager** and **Kernel Manager**, which would then allow some introduction as to the importance the `jupyter_client` package serves within the Server architecture.

Author: That makes sense, thanks @kevin-bates!
